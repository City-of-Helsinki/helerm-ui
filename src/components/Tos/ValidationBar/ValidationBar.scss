.missing-attributes {
  margin-bottom: 25px;
}

.missing-attribute {
  color: red;
}

.sidebar-content {
  h4 {
    margin-bottom: 20px;
  }

  h5 {
    margin: 20px 0 10px;
  }

  .element-name {
    margin-bottom: 5px;
  }
}
