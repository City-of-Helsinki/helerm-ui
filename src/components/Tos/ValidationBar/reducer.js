import update from 'immutability-helper';
import { createAction, handleActions } from 'redux-actions';

const initialState = {
  is_open: false
};

export const SET_VALIDATION_VISIBILITY = 'setValidationVisibilityAction';

export function setValidationVisibility (value) {
  return createAction(SET_VALIDATION_VISIBILITY)(value);
}

const setValidationVisibilityAction = (state, { payload }) => {
  return update(state, {
    is_open: { $set: payload }
  });
};

export default handleActions({
  setValidationVisibilityAction
}, initialState);
