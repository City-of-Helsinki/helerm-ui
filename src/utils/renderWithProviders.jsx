/* eslint-disable react/prop-types */
import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import storeCreator from '../store/createStore';

const renderWithProviders = (
  ui,
  { preloadedState = {}, history, store = storeCreator(history, preloadedState), ...renderOptions } = {},
) => {
  const Wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};


export default renderWithProviders;

