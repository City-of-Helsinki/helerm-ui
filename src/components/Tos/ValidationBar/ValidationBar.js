import React, { Component } from 'react';
import SideBar from 'react-sidebar';
import map from 'lodash/map';
import './ValidationBar.scss';

import { validateTOS, validateRecord } from '../../../utils/validators';

const styles = {
  sidebar: {
    width: 256,
    height: '100%',
    padding: '10px 0 0 20px',
    backgroundColor: 'white',
    zIndex: '99'
  },
  sidebarLink: {
    display: 'block',
    color: '#757575',
    textDecoration: 'none'
  },
  divider: {
    margin: '8px 0',
    height: 1,
    backgroundColor: '#757575'
  },
  content: {
    height: '100%'
  }
};

export class ValidationBar extends Component {
  constructor (props) {
    super(props);
    this.renderContent = this.renderContent.bind(this);
  }

  generateInvalidAttributes (validate, values) {
    const mappedInvalidAttributes = map(validate(values, this.props.attributeTypes), (item, index) => (
      <div key={index} className='missing-attribute'>
        {'• '}{item}
      </div>
    ));

    return (
      <div className='missing-attributes'>
        {mappedInvalidAttributes}
      </div>
    );
  }

  generateAttributeSection (validate, element, index) {
    const invalidAttributes = validate(element, this.props.attributeTypes);

    if (invalidAttributes.length) {
      return (
        <div key={index}>
          <div className='parent-name'>{element.name}</div>
          <div className='missing-attributes'>
            {map(invalidAttributes, (item) => (
              <div className='missing-attribute'>
                {'• '}{item}
              </div>
            ))}
          </div>
        </div>
      );
    }
  }

  renderContent () {
    const { selectedTOS } = this.props;

    const invalidTOSAttributes = this.generateInvalidAttributes(validateTOS, selectedTOS);
    const invalidPhaseAttributes = [];
    const invalidActionAttributes = [];
    const invalidRecordAttributes = map(selectedTOS.records, (record, index) => (
        this.generateAttributeSection(validateRecord, record, index)
      )
    );

    return (
      <div className='sidebar-content'>
        <h4>Puuttuvat metatiedot</h4>
        {invalidTOSAttributes.length !== 0 &&
          <h5>TOS-metatiedot</h5>}
        {invalidTOSAttributes}
        {invalidPhaseAttributes.length !== 0 &&
          <h5>Käsittelyvaiheet</h5>}
        {invalidPhaseAttributes}
        {invalidActionAttributes.length !== 0 &&
          <h5>Toimenpiteet</h5>}
        {invalidActionAttributes}
        {invalidRecordAttributes.length !== 0 &&
          <h5>Asiakirjat</h5>}
        {invalidRecordAttributes}
      </div>
    );
  }

  render () {
    const { children, selectedTOS } = this.props;
    const sidebarContent = selectedTOS.id ? this.renderContent() : <div/>;

    return (
      <div>
        <SideBar
          sidebar={sidebarContent}
          open={this.props.is_open}
          onSetOpen={() => this.props.setValidationVisibility(false)}
          pullRight={true}
          styles={styles}>
          {children}
        </SideBar>
      </div>
    );
  }
};

ValidationBar.propTypes = {
  attributeTypes: React.PropTypes.object.isRequired,
  children: React.PropTypes.array.isRequired,
  is_open: React.PropTypes.bool.isRequired,
  selectedTOS: React.PropTypes.object.isRequired,
  setValidationVisibility: React.PropTypes.func.isRequired
};

export default ValidationBar;
