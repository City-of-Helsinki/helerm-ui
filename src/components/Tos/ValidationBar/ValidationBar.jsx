/* eslint-disable react/forbid-prop-types */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { map, find } from 'lodash';

import './ValidationBar.scss';

import {
  validateTOS,
  validatePhase,
  validateAction,
  validateRecord,
  validateTOSWarnings,
  validatePhaseWarnings,
  validateActionWarnings,
  validateRecordWarnings,
} from '../../../utils/validators';
import { VALIDATION_FILTER_ERROR, VALIDATION_FILTER_WARN } from '../../../constants';

const ATTRIBUTE_NAME_FIELDS = ['PhaseType', 'ActionType', 'TypeSpecifier'];

const ValidationBar = (props) => {
  const { selectedTOS, setValidationVisibility, top, attributeTypes, scrollToMetadata, scrollToType } = props;

  const [filter, setFilter] = useState('');

  const getFilterByStatus = (value) => (!filter ? true : value === filter);

  const generateInvalidTosAttributes = (validate, validateWarn, values) => {
    const showInvalidAttributes = getFilterByStatus(VALIDATION_FILTER_ERROR);
    const showWarnAttributes = getFilterByStatus(VALIDATION_FILTER_WARN);
    const invalidAttributes = showInvalidAttributes ? validate(values, attributeTypes) : [];
    const warnAttributes = showWarnAttributes ? validateWarn(values, attributeTypes) : [];
    const mappedInvalidAttributes = map(invalidAttributes, (item, index) => (
      <div key={index} className='missing-attribute'>
        <i className='fa-solid fa-circle-xmark' /> {attributeTypes[item].name}
      </div>
    ));
    const mappedWarnAttributes = map(warnAttributes, (item, index) => (
      <div key={index} className='warn-attribute'>
        <i className='fa-solid fa-circle-exclamation' /> {attributeTypes[item].name}
      </div>
    ));

    if (invalidAttributes.length || warnAttributes.length) {
      return (
        <div className='missing-attributes'>
          {mappedInvalidAttributes}
          {mappedWarnAttributes}
        </div>
      );
    }

    return null;
  };

  const getInvalidSection = (type, section, validateRequired, validateWarn, children) => {
    const hasChildren = children !== undefined && children.filter((x) => x !== null).length > 0;

    if (!section || (type !== 'record' && !hasChildren)) {
      return null;
    }

    const invalidAttributes = validateRequired ? validateRequired(section, attributeTypes) : [];
    const warnAttributes = validateWarn ? validateWarn(section, attributeTypes) : [];
    const nameAttribute = section.attributes ? find(ATTRIBUTE_NAME_FIELDS, (field) => !!section.attributes[field]) : '';

    if (invalidAttributes.length || warnAttributes.length || children?.length) {
      return (
        <div className={`sidebar-content-${type}`} key={section.id}>
          <button
            type='button'
            className='parent-name'
            onClick={() => scrollToType(type, section.id)}
            onKeyUp={(e) => {
              if (e.key === 'Enter') {
                scrollToType(type, section.id);
              }
            }}
          >
            {nameAttribute ? section.attributes[nameAttribute] : ''}
          </button>
          <div className='missing-attributes'>
            {map(invalidAttributes, (item, key) => (
              <div key={key} className='missing-attribute'>
                <i className='fa-solid fa-triangle-exclamation' /> {attributeTypes[item].name}
              </div>
            ))}
            {map(warnAttributes, (item, key) => (
              <div key={key} className='warn-attribute'>
                <i className='fa-solid fa-circle-exclamation' /> {attributeTypes[item].name}
              </div>
            ))}
          </div>
          {children}
        </div>
      );
    }

    return null;
  };

  // eslint-disable-next-line sonarjs/cognitive-complexity
  const generateInvalidAttributes = () => {
    const showInvalidAttributes = getFilterByStatus(VALIDATION_FILTER_ERROR);
    const showWarnAttributes = getFilterByStatus(VALIDATION_FILTER_WARN);

    // get invalid sections of phase-action-record -tree as div elements
    const invalidPhases = Object.values(selectedTOS.phases)
      .map((phase) => ({
        id: phase.id,
        invalidSection: getInvalidSection(
          'phase',
          phase,
          showInvalidAttributes ? validatePhase : null,
          showWarnAttributes ? validatePhaseWarnings : null,
          phase.actions
            .map((actionId) => selectedTOS.actions?.[actionId])
            .map((action) =>
              getInvalidSection(
                'action',
                action,
                showInvalidAttributes ? validateAction : null,
                showWarnAttributes ? validateActionWarnings : null,
                action?.records
                  .map((recordId) => selectedTOS.records?.[recordId])
                  .map((record) =>
                    getInvalidSection(
                      'record',
                      record,
                      showInvalidAttributes ? validateRecord : null,
                      showWarnAttributes ? validateRecordWarnings : null,
                    ),
                  ),
              ),
            ),
        ),
      }))
      .filter(({ invalidSection }) => Boolean(invalidSection))
      .map(({ id, invalidSection }) => <div key={id}>{invalidSection}</div>);

    return invalidPhases.length > 0 ? invalidPhases : null;
  };

  const renderInvalidContent = () => {
    const invalidTOSAttributes = generateInvalidTosAttributes(validateTOS, validateTOSWarnings, selectedTOS);
    const invalidAttributes = generateInvalidAttributes();

    if (invalidTOSAttributes || invalidAttributes?.length > 0) {
      return (
        <div className='sidebar-invalid-content'>
          {invalidTOSAttributes && (
            <button
              className='scroll-to-button'
              type='button'
              onClick={scrollToMetadata}
              onKeyUp={(e) => {
                if (e.key === 'Enter') {
                  scrollToMetadata();
                }
              }}
            >
              KÃ¤sittelyprosessi
            </button>
          )}
          {invalidTOSAttributes}
          {invalidAttributes}
        </div>
      );
    }

    return (
      <div className='no-missing-attributes'>
        <div className='fa-solid fa-circle-check' />
      </div>
    );
  };

  return (
    <div className='validation-bar' style={{ top: top + 47 }}>
      <div className='validation-bar-header'>
        <h3>
          Esitarkastus{' '}
          <button
            type='button'
            className='sidebar-close-button pull-right'
            onClick={() => setValidationVisibility(false)}
            aria-label='Esitarkastus'
          >
            <i className='fa-solid fa-xmark' />
          </button>
        </h3>
        <div className='sidebar-filter'>
          <button
            type='button'
            className={classnames('sidebar-filter-all btn btn-sm', { 'btn-default': filter === '' })}
            onClick={() => setFilter('')}
          >
            Kaikki
          </button>
          <button
            type='button'
            className={classnames('sidebar-filter-warn btn btn-sm', {
              'btn-default': filter === VALIDATION_FILTER_WARN,
            })}
            onClick={() => setFilter(VALIDATION_FILTER_WARN)}
          >
            <i className='fa-solid fa-circle-exclamation' /> Huomautukset
          </button>
          <button
            type='button'
            className={classnames('sidebar-filter-error btn btn-sm', {
              'btn-default': filter === VALIDATION_FILTER_ERROR,
            })}
            onClick={() => setFilter(VALIDATION_FILTER_ERROR)}
          >
            <i className='fa-solid fa-triangle-exclamation' /> Virheet
          </button>
        </div>
      </div>
      {selectedTOS.id ? <div className='sidebar-content'>{renderInvalidContent()}</div> : <div />}
    </div>
  );
};

ValidationBar.propTypes = {
  attributeTypes: PropTypes.object.isRequired,
  scrollToMetadata: PropTypes.func.isRequired,
  scrollToType: PropTypes.func.isRequired,
  selectedTOS: PropTypes.object.isRequired,
  setValidationVisibility: PropTypes.func.isRequired,
  top: PropTypes.number.isRequired,
};

export default ValidationBar;
