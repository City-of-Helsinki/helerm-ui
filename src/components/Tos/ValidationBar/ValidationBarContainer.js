// import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import ValidationBar from './ValidationBar';

const mapStateToProps = (state) => {
  const { selectedTOS } = state;

  return {
    selectedTOS
  };
};

export default connect(mapStateToProps)(ValidationBar);
