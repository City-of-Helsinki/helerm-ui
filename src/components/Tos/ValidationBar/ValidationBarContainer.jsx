import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import ValidationBar from './ValidationBar';
import { setValidationVisibility } from './reducer';

const mapDispatchToProps = (dispatch) => ({
  setValidationVisibility: bindActionCreators(setValidationVisibility, dispatch),
});

const mapStateToProps = (state) => {
  const { ui, selectedTOS } = state;

  return {
    attributeTypes: ui.attributeTypes,
    selectedTOS,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ValidationBar);
