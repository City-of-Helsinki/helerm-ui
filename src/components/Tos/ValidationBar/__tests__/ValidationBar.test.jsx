import React from 'react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import renderWithProviders from '../../../../utils/renderWithProviders';
import attributeRules from '../../../../utils/mocks/attributeRules.json';
import errorsAndWarningsTOS from '../../../../utils/mocks/errorsAndWarningsTOS.json';
import validTOS from '../../../../utils/mocks/validTOS.json';
import ValidationBar from '../ValidationBar';

const baseMocks = {
  setValidationVisibility: vi.fn(),
  selectedTOS: errorsAndWarningsTOS,
};

const renderComponent = (history, mocks = baseMocks) =>
  renderWithProviders(
    <Router history={history}>
      <ValidationBar
        selectedTOS={mocks.selectedTOS}
        attributeTypes={attributeRules}
        scrollToMetadata={vi.fn()}
        scrollToType={vi.fn()}
        top={0}
        setValidationVisibility={mocks.setValidationVisibility}
      />
    </Router>,
    {
      history,
    },
  );

describe('<Navigation />', () => {
  it('should render correctly', () => {
    const history = createBrowserHistory();

    const { container } = renderComponent(history);

    expect(container).toMatchSnapshot();
  });

  it('should change filter', async () => {
    const history = createBrowserHistory();

    renderComponent(history);

    const allFilter = screen.getByRole('button', { name: 'Kaikki' });
    const warningsFilter = screen.getByRole('button', { name: 'Huomautukset' });
    const errorsFilter = screen.getByRole('button', { name: 'Virheet' });

    const user = userEvent.setup();

    await user.click(warningsFilter);

    const activeClassName = 'btn-default';

    expect(warningsFilter).toHaveClass(activeClassName);
    expect(allFilter).not.toHaveClass(activeClassName);
    expect(errorsFilter).not.toHaveClass(activeClassName);

    await user.click(allFilter);

    expect(allFilter).toHaveClass(activeClassName);
    expect(warningsFilter).not.toHaveClass(activeClassName);
    expect(errorsFilter).not.toHaveClass(activeClassName);

    await user.click(errorsFilter);

    expect(errorsFilter).toHaveClass(activeClassName);
    expect(allFilter).not.toHaveClass(activeClassName);
    expect(warningsFilter).not.toHaveClass(activeClassName);

    await user.click(allFilter);

    expect(allFilter).toHaveClass(activeClassName);
    expect(warningsFilter).not.toHaveClass(activeClassName);
    expect(errorsFilter).not.toHaveClass(activeClassName);
  });

  it('should close validation bar', async () => {
    const history = createBrowserHistory();

    const mockSetValidationVisibility = vi.fn();
    const mocks = { ...baseMocks, setValidationVisibility: mockSetValidationVisibility };

    renderComponent(history, mocks);

    const closeButton = screen.getAllByRole('button')[0];

    const user = userEvent.setup();

    await user.click(closeButton);

    expect(mockSetValidationVisibility).toHaveBeenCalled();
  });

  it('should render errors for invalid TOS', () => {
    const history = createBrowserHistory();

    renderComponent(history);
    // find error group
    expect(screen.getByRole('heading', { name: 'Käsittelyprosessi' })).toBeInTheDocument();
    // find correct errors
    expect(screen.getByText('Julkisuusluokka')).toBeInTheDocument();
    expect(screen.getByText('Säilytysajan laskentaperuste')).toBeInTheDocument();
    expect(screen.getByText('Rekisteröinti/ Tietojärjestelmä')).toBeInTheDocument();
  });

  it('should not render errors for valid TOS', () => {
    const history = createBrowserHistory();

    const mocks = { ...baseMocks, selectedTOS: validTOS };

    renderComponent(history, mocks);

    expect(screen.queryByRole('heading', { name: 'Käsittelyprosessi' })).not.toBeInTheDocument();
  });
});
