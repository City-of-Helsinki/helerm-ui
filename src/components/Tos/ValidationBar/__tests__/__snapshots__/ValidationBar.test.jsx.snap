// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> should render correctly 1`] = `
<div>
  <div
    class="validation-bar"
    style="top: 47px;"
  >
    <div
      class="validation-bar-header"
    >
      <h3>
        Esitarkastus
         
        <button
          class="sidebar-close-button pull-right"
          type="button"
        >
          <i
            class="fa-solid fa-xmark"
          />
        </button>
      </h3>
      <div
        class="sidebar-filter"
      >
        <button
          class="sidebar-filter-all btn btn-sm btn-default"
          type="button"
        >
          Kaikki
        </button>
        <button
          class="sidebar-filter-warn btn btn-sm"
          type="button"
        >
          <i
            class="fa-solid fa-circle-exclamation"
          />
           Huomautukset
        </button>
        <button
          class="sidebar-filter-error btn btn-sm"
          type="button"
        >
          <i
            class="fa-solid fa-triangle-exclamation"
          />
           Virheet
        </button>
      </div>
    </div>
    <div
      class="sidebar-content"
    >
      <div
        class="sidebar-invalid-content"
      >
        <h5>
          Käsittelyprosessi
        </h5>
        <div
          class="missing-attributes"
        >
          <div
            class="missing-attribute"
          >
            <i
              class="fa-solid fa-circle-xmark"
            />
             
            Julkisuusluokka
          </div>
          <div
            class="missing-attribute"
          >
            <i
              class="fa-solid fa-circle-xmark"
            />
             
            Säilytysajan laskentaperuste
          </div>
          <div
            class="warn-attribute"
          >
            <i
              class="fa-solid fa-circle-exclamation"
            />
             
            Rekisteröinti/ Tietojärjestelmä
          </div>
        </div>
        <div>
          <div
            class="sidebar-content-phase"
          >
            <div
              class="parent-name"
            >
              Vireillepano/-tulo
            </div>
            <div
              class="missing-attributes"
            >
              <div
                class="warn-attribute"
              >
                <i
                  class="fa-solid fa-circle-exclamation"
                />
                 
                Rekisteröinti/ Tietojärjestelmä
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="sidebar-content-phase"
          >
            <div
              class="parent-name"
            >
              Vireillepano/-tulo
            </div>
            <div
              class="missing-attributes"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
